optuna/visualization/_pareto_front.py

```diff
@@ -251,34 +251,47 @@ def _get_pareto_front_info(
             "Use either `targets` or `axis_order`."
         )
 
+    feasible_trials = []
+    infeasible_trials = []
+    has_constraints = False
     if constraints_func is not None:
-        warn_experimental_argument("constraints_func")
-        feasible_trials = []
-        infeasible_trials = []
+        warnings.warn(
+            "`constraints_func` has been deprecated in v4.0.0. "
+            "This feature will be removed in v6.0.0. "
+            "See https://github.com/optuna/optuna/releases/tag/v4.0.0.",
+            FutureWarning,
+        )
         for trial in study.get_trials(deepcopy=False, states=(TrialState.COMPLETE,)):
             if all(map(lambda x: x <= 0.0, constraints_func(trial))):
                 feasible_trials.append(trial)
             else:
                 infeasible_trials.append(trial)
-        best_trials = _get_pareto_front_trials_by_trials(feasible_trials, study.directions)
-        if include_dominated_trials:
-            non_best_trials = _get_non_pareto_front_trials(feasible_trials, best_trials)
-        else:
-            non_best_trials = []
+        has_constraints = True
+    else:
+        for trial in study.get_trials(deepcopy=False, states=(TrialState.COMPLETE,)):
+            constraints = trial.system_attrs.get(_CONSTRAINTS_KEY)
+
+            if constraints is not None:
+                has_constraints = True
 
-        if len(best_trials) == 0:
-            _logger.warning("Your study does not have any completed and feasible trials.")
+            if constraints is None or all([x <= 0.0 for x in constraints]):
+                feasible_trials.append(trial)
+            else:
+                infeasible_trials.append(trial)
```

Comment:
Thank you for your suggestion!
In your suggestion, deprecated parts are separated into several parts apart (i.e., parts including `constraints_func` are not gathered), so it makes it rather a bit more difficult to remove deprecated parts in the future, IMO. I think it would be easier to understand the logic if deprecated and non-deprecated parts are branched off from a earlier stage, rather than having branches within `for` loop.
What do you think?

Reply:
Fair enough!
Then what about the following?

```python
    feasible_trials = []
    infeasible_trials = []
    has_constraint = False
    for trial in study.get_trials(deepcopy=False, states=(TrialState.COMPLETE,)):
        if constraints_func is not None:
            has_constraint = True
            if all(map(lambda x: x <= 0.0, constraints_func(trial))):
                feasible_trials.append(trial)
            else:
                infeasible_trials.append(trial)
            continue
        
        constraints = trial.system_attrs.get(_CONSTRAINTS_KEY)
        has_constraint |= bool(constraints is not None)
        if constraints is not None and all(x <= 0.0 for x in constraints):
            feasible_trials.append(trial)
        else:
            infeasible_trials.append(trial)
```